firmware/src/moving_light.c

```diff
@@ -0,0 +1,356 @@
+    return RDMResponder_BuildNack(header, NR_SUB_DEVICE_OUT_OF_RANGE);
+  }
+
+  if (header->command_class == DISCOVERY_COMMAND) {
+    return RDMResponder_HandleDiscovery(header, param_data);
+  }
+
+  // This model has no sub devices.
+  if (header->command_class == GET_COMMAND && sub_device == SUBDEVICE_ALL) {
+    return RDMResponder_BuildNack(header, NR_SUB_DEVICE_OUT_OF_RANGE);
+  }
+
+  return RDMResponder_DispatchPID(header, param_data);
+}
+
+static void MovingLightModel_Tasks() {
+  if (g_moving_light.lamp_state == LAMP_STRIKE &&
+      CoarseTimer_HasElapsed(g_moving_light.lamp_strike_time,
+                             LAMP_STRIKE_DELAY)) {
+    g_moving_light.lamp_state = LAMP_ON;
+    g_moving_light.lamp_strikes++;
+  }
+}
+
+const ModelEntry MOVING_LIGHT_MODEL_ENTRY = {
+  .model_id = MOVING_LIGHT_MODEL_ID,
+  .activate_fn = MovingLightModel_Activate,
+  .deactivate_fn = MovingLightModel_Deactivate,
+  .ioctl_fn = MovingLightModel_Ioctl,
+  .request_fn = MovingLightModel_HandleRequest,
+  .tasks_fn = MovingLightModel_Tasks
+};
+
+static const PIDDescriptor PID_DESCRIPTORS[] = {
+  {PID_SUPPORTED_PARAMETERS, RDMResponder_GetSupportedParameters,
+    (PIDCommandHandler) NULL},
+  {PID_DEVICE_INFO, RDMResponder_GetDeviceInfo, (PIDCommandHandler) NULL},
+  {PID_PRODUCT_DETAIL_ID_LIST, RDMResponder_GetProductDetailIds,
+    (PIDCommandHandler) NULL},
+  {PID_DEVICE_MODEL_DESCRIPTION, RDMResponder_GetDeviceModelDescription,
+    (PIDCommandHandler) NULL},
+  {PID_MANUFACTURER_LABEL, RDMResponder_GetManufacturerLabel,
+    (PIDCommandHandler) NULL},
+  {PID_DEVICE_LABEL, RDMResponder_GetDeviceLabel, RDMResponder_SetDeviceLabel},
+  {PID_SOFTWARE_VERSION_LABEL, RDMResponder_GetSoftwareVersionLabel,
+    (PIDCommandHandler) NULL},
+  {PID_DEVICE_HOURS, MovingLightModel_GetUInt32, MovingLightModel_SetUInt32},
+  {PID_LAMP_HOURS, MovingLightModel_GetUInt32, MovingLightModel_SetUInt32},
+  {PID_LAMP_STRIKES, MovingLightModel_GetUInt32, MovingLightModel_SetUInt32},
+  {PID_LAMP_STATE, MovingLightModel_GetUInt8, MovingLightModel_SetLampState},
+  {PID_LAMP_ON_MODE, MovingLightModel_GetUInt8, MovingLightModel_SetLampOnMode},
+  {PID_DEVICE_POWER_CYCLES, MovingLightModel_GetUInt32,
+    MovingLightModel_SetUInt32},
+  {PID_DISPLAY_INVERT, MovingLightModel_GetUInt8,
+    MovingLightModel_SetDisplayInvert},
+  {PID_DISPLAY_LEVEL, MovingLightModel_GetUInt8, MovingLightModel_SetUInt8},
+  {PID_PAN_INVERT, MovingLightModel_GetBool, MovingLightModel_SetBool},
+  {PID_TILT_INVERT, MovingLightModel_GetBool, MovingLightModel_SetBool},
+  {PID_PAN_TILT_SWAP, MovingLightModel_GetBool, MovingLightModel_SetBool},
+  {PID_IDENTIFY_DEVICE, RDMResponder_GetIdentifyDevice,
+    RDMResponder_SetIdentifyDevice},
+  {PID_POWER_STATE, MovingLightModel_GetUInt8, MovingLightModel_SetPowerState}
+};
+
+static const ProductDetailIds PRODUCT_DETAIL_ID_LIST = {
+  .ids = {PRODUCT_DETAIL_TEST, PRODUCT_DETAIL_CHANGEOVER_MANUAL,
+          PRODUCT_DETAIL_LED},
```

Comment:
I'm not sure you'd implement the LAMP_STATE PID (or at least the set side) unless you had a lamp that needed striking though, it seems redundant otherwise.